upcloud/error.go

```diff
@@ -3,35 +3,151 @@ package upcloud
 import (
 	"encoding/json"
 	"fmt"
+	"strings"
+
+	"github.com/UpCloudLtd/upcloud-go-api/v5/upcloud/client"
 )
 
+func NewError(clientErr *client.Error) error {
+	if clientErr == nil {
+		return nil
+	}
+
+	ucErr := &Error{}
+
+	err := json.Unmarshal(clientErr.ResponseBody, ucErr)
+	if err != nil {
+		return fmt.Errorf("received malformed client error: %s", err)
+	}
+
+	ucErr.Status = clientErr.ErrorCode
+	ucErr.clientError = clientErr
+
+	return ucErr
+}
+
 // Error represents an error
 type Error struct {
-	ErrorCode    string `json:"error_code"`
-	ErrorMessage string `json:"error_message"`
+	// Code is a short, programmatic identifier of an error
+	Code string `json:"code"`
 
-	// HTTP Status code
+	// Message is a human-readable description of an error
+	Message string `json:"message"`
+
+	// Status is an HTTP Status code from the response
 	Status int `json:"-"`
+
+	clientError  *client.Error
+	problemError *ProblemError
+}
+
+// IsProblem checks if the `Error` has additional information on the problem with HTTP request (conforming to RFC7807)
+func (err *Error) IsProblem() bool {
+	return err.clientError.Type == client.ErrorTypeProblem
+}
+
+// Problem returns additional information on the problem with HTTP request (conforming to RFC7807), assuming they were included in error response.
+// Second return value is a boolean that will be true if the additional information actually exists.
+func (err *Error) Problem() (*ProblemError, bool) {
+	if err.IsProblem() {
+		return err.problemError, true
+	}
+
+	return nil, false
+}
+
+// Error implements the Error interface
+func (err *Error) Error() string {
+	if err.IsProblem() {
+		return err.problemError.Error()
+	}
+
+	return fmt.Sprintf("%s (%s)", err.Message, err.Code)
 }
 
-// UnmarshalJSON is a custom unmarshaller that deals with
-// deeply embedded values.
-func (e *Error) UnmarshalJSON(b []byte) error {
-	type localError Error
-	v := struct {
-		Error localError `json:"error"`
+func (err *Error) UnmarshalJSON(buf []byte) error {
+	localError := struct {
+		LegacyError *struct {
+			ErrorCode    string `json:"error_code,omitempty"`
+			ErrorMessage string `json:"error_message,omitempty"`
+		} `json:"error,omitempty"`
+		ProblemType          string                     `json:"type,omitempty"`
+		ProblemTitle         string                     `json:"title,omitempty"`
+		ProblemCorrelationID string                     `json:"correlation_id,omitempty"`
+		ProblemStatus        int                        `json:"status,omitempty"`
+		ProblemInvalidParams []ProblemErrorInvalidParam `json:"invalid_params,omitempty"`
 	}{}
```

Comment:
This is a bit hacky. It relies on the `omitempty` json tag to check what we actually got from the API.

[Here is an alternative implementation of this custom unmarshaler](https://github.com/UpCloudLtd/upcloud-go-api/pull/209/files). The logic looks nicer, but the overall solution is convoluted and more limiting IMO.